tests/test_black.py

```diff
@@ -153,6 +154,27 @@ def test_empty_ff(self) -> None:
             os.unlink(tmp_file)
         self.assertFormatEqual(expected, actual)
 
+    @patch("black.dump_to_file", dump_to_stderr)
+    def test_one_empty_line(self) -> None:
+        for nl in ["\n", "\r\n"]:
+            source = expected = nl
+            actual = fs(source)
+            self.assertFormatEqual(expected, actual)
+            black.assert_equivalent(source, actual)
+            black.assert_stable(source, actual, DEFAULT_MODE)
+
+    def test_one_empty_line_ff(self) -> None:
+        for nl in ["\n", "\r\n"]:
+            expected = nl
+            tmp_file = Path(black.dump_to_file(nl))
+            try:
+                self.assertFalse(ff(tmp_file, write_back=black.WriteBack.YES))
+                with open(tmp_file, "rb") as f:
+                    actual = f.read().decode("utf8")
+            finally:
+                os.unlink(tmp_file)
+            self.assertFormatEqual(expected, actual)
```

Comment:
The tests failing on Windows seem to confirm that writing a newline actually writes a `\r\n` instead of `\n`.

```
Expected tree:
file_input
  ENDMARKER '\n' ''
/file_input
Actual tree:
file_input
  ENDMARKER '\r\n' ''
/file_input
```

I'm getting this on Windows:

```python
>>> with open("temp.py", "w") as f:
...     f.write("\n")
...
1
>>>
>>> with open("temp.py", "rb") as f:
...     print(f.read())
...
b'\r\n'                
```

Reply:
The issue goes away if the file is written in binary mode:

```python
>>> with open("temp.py", "wb") as f:
...     f.write(b"\n")
...
1
>>> with open("temp.py", "rb") as f:
...     print(f.read())
...
b'\n'                
```

I'll keep investigating. This may be affecting other tests without us noticing, because behavior on Windows might be tested incorrectly

Reply:
Ok, so if I change `dump_to_file` to writing files in binary mode, all the tests are passed (for Ubuntu, Windows, and macOS). The downside is that in Windows, it might introduce an unexpected behavior, because under normal conditions we expect `dump_to_file` to write files using the correct newline (`\r\n`). I'm not sure if this is a bug.

Another solution could be write the file directly in the test that makes the "bug" visible, but might be non-intuitive. In the meantime, I'll remove the commit that tested this. This is the change that I applied (described in the first paragraph):

```diff
diff --git a/src/black/output.py b/src/black/output.py
index f4c17f28ea..19a178e89d 100644
--- a/src/black/output.py
+++ b/src/black/output.py
@@ -96,10 +96,10 @@ def color_diff(contents: str) -> str:
 def dump_to_file(*output: str, ensure_final_newline: bool = True) -> str:
     """Dump `output` to a temporary file. Return path to the file."""
     with tempfile.NamedTemporaryFile(
-        mode="w", prefix="blk_", suffix=".log", delete=False, encoding="utf8"
+        mode="wb", prefix="blk_", suffix=".log", delete=False
     ) as f:
         for lines in output:
-            f.write(lines)
+            f.write(lines.encode("utf8"))
             if ensure_final_newline and lines and lines[-1] != "\n":
-                f.write("\n")
+                f.write("\n".encode("utf8"))
     return f.name
```

Reply:
Patched the test by explicitly rewriting the temporary file in binary mode on Windows only:

```diff
diff --git a/tests/test_black.py b/tests/test_black.py
index b49830f..9ae17af 100644
--- a/tests/test_black.py
+++ b/tests/test_black.py
@@ -167,6 +167,12 @@ class BlackTestCase(BlackBaseTestCase):
         for nl in ["\n", "\r\n"]:
             expected = nl
             tmp_file = Path(black.dump_to_file(nl))
+            if system() == "Windows":
+                # Writing files in text mode automatically uses the system newline,
+                # but in this case we don't want this for testing reasons. See:
+                # https://github.com/psf/black/pull/3348
+                with open(tmp_file, "wb") as f:
+                    f.write(nl.encode("utf-8"))
             try:
                 self.assertFalse(ff(tmp_file, write_back=black.WriteBack.YES))
                 with open(tmp_file, "rb") as f:
```